src/Common/FileSegment.cpp

```diff
@@ -593,10 +500,62 @@ void FileSegment::completeImpl(std::lock_guard<std::mutex> & cache_lock, std::lo
         }
     }
 
-    if (!downloader_id.empty() && (isDownloaderImpl(segment_lock) || is_last_holder))
+    switch (download_state)
     {
-        LOG_TEST(log, "Clearing downloader id: {}, current state: {}", downloader_id, stateToString(download_state));
-        downloader_id.clear();
+        case State::SKIP_CACHE:
+        {
+            if (is_last_holder)
+                cache->remove(key(), offset(), cache_lock, segment_lock);
+
+            return;
+        }
+        case State::DOWNLOADED:
+        {
+            assert(downloaded_size == range().size());
+            assert(is_downloaded);
+            break;
+        }
+        case State::DOWNLOADING:
+        case State::EMPTY:
+        {
+            assert(!is_last_holder);
+            break;
+        }
+        case State::PARTIALLY_DOWNLOADED:
+        case State::PARTIALLY_DOWNLOADED_NO_CONTINUATION:
+        {
+            if (is_last_holder)
```

Comment:
> So if we are not the last holder than last holder will do it?

Yes, because `completeWithoutState` is always called for any user of the file segment - from destructor of `FileSegmentsHolder` -- file segments are always put in a special holder which calls complete() for the in destructor. E.g. eventually `is_last_holder` will be `true`.